ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java

```diff
@@ -133,107 +222,245 @@ long getMaxBlobSize() {
+            // delete is not relevant yet and it's in the same index segment as the original put. We need to find the
+            // original put since it's still valid
+            IndexValue originalPutValue = getPutRecordForDeletedKey(indexEntry.getKey(), indexValue);
+            validIndexEntries.add(new IndexEntry(indexEntry.getKey(), originalPutValue));
+          }
         }
-      } else if (!isExpired(indexValue.getExpiresAtMs(), expirationReferenceTimeInMs) && !deletedKeys.contains(
-          indexEntry.getKey())) {
-        // put record that is not deleted and not expired according to deleteReferenceTimeInMs and expirationReferenceTimeInMs
-        validIndexValues.add(indexValue);
+      } else if (!isExpired(indexValue.getExpiresAtMs(), referenceTimeInMs) && (
+          !deletedKeys.containsKey(indexEntry.getKey()) || deletedKeys.get(indexEntry.getKey()) >= referenceTimeInMs)) {
+        // put record that is not deleted and not expired according to given reference times
+        validIndexEntries.add(indexEntry);
       }
     }
-    return validIndexValues;
+    return validIndexEntries;
+  }
+
+  /**
+   * Given {@link ScanResults} and a {@link TimeRange}, try to find the latest point in time that is within the
+   * {@link TimeRange} and the log segment forecast range.
+   * @param results the {@link ScanResults} with the log segment forecast range information
+   * @param timeRange the {@link TimeRange} used to find the latest shared point in time with the log segment forecast
+   *                  range
+   * @return the latest shared point in time in milliseconds if there is one, otherwise REF_TIME_OUT_OF_BOUNDS (-1)
+   * is returned
+   */
+  private long getLogSegmentRefTimeMs(ScanResults results, TimeRange timeRange) {
+    long refTimeInMs = results == null || timeRange.getStartTimeInMs() >= results.logSegmentForecastEndTimeMs
+        || timeRange.getEndTimeInMs() < results.logSegmentForecastStartTimeMs ? REF_TIME_OUT_OF_BOUNDS
+        : timeRange.getEndTimeInMs();
+    if (refTimeInMs != REF_TIME_OUT_OF_BOUNDS && refTimeInMs >= results.logSegmentForecastEndTimeMs) {
+      refTimeInMs = results.logSegmentLastBucketTimeMs;
+    }
+    return refTimeInMs;
+  }
+
+  /**
+   * Find the original PUT that was deleted within the same index segment.
+   * @param key the {@link StoreKey} for the entry
+   * @param deleteIndexValue the {@link IndexValue} of the delete
+   * @return a copy of the original put {@link IndexValue}
+   * @throws StoreException
+   */
+  private IndexValue getPutRecordForDeletedKey(StoreKey key, IndexValue deleteIndexValue) throws StoreException {
+    BlobReadOptions originalPut = index.getBlobReadInfo(key, EnumSet.of(StoreGetOptions.Store_Include_Deleted));
+    Offset originalPutOffset = new Offset(originalPut.getLogSegmentName(), originalPut.getOffset());
+    return new IndexValue(originalPut.getSize(), originalPutOffset, originalPut.getExpiresAtMs(), Utils.Infinite_Time,
```

Comment:
for op time, how about using the optime in the delete? This is only called when the put and delete are in the same segment and the optime of the delete is a good approximation (rather than `Utils.Infinite_Time`)